### PR TITLE
TINY-7478: Added table cell border style and width toolbar buttons/menu items

### PR DESCRIPTION
**Pull request by HAFRMO**
Related Ticket:

Description of Changes:
Added cell border and style

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/plugins/table/main/ts/api/Settings.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/api/Settings.ts:46]:
```suggestion
const defaultCellBorderWidth = Arr.range(5, (i) => {
  const size = `${i + 1}px`;
  return { title: size, value: size };
});
```

File: modules/tinymce/src/plugins/table/main/ts/api/Settings.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/api/Settings.ts:65]:
```suggestion
const defaultCellBorderStyles = Arr.map([ 'Solid', 'Dashed', 'Dotted', 'Double' ], (type) => {
  return { title: type, value: type.toLowerCase() };
});
```

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderStyleTest.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderStyleTest.ts:39]: This is never used in this test, so why are we logging the events? I'm not sure if this should be removed or if we should be asserting the events in the tests.

- [ ] Remove or add an assertion

File: modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts:254]: This is now being duplicated 6 times with just a different value (3 in each file). Can you please abstract that out into a helper function?

- [ ] Clean up duplication

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderStyleTest.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderStyleTest.ts:41]: These setup content and selection all seem to be duplicated exactly between your different tests. Please create a test helper file in `test/ts/module/test/` and move them there to remove the duplication.

I believe these are the ones that are the same across your previous PRs and this PR:
- `setEditorContentTableSingleCellAndSelection`
- `assertSingleCellStructureIsStyleless`
- `setEditorContentTableMultipleCellsAndSelection`
- `assertMultipleCellsStructureIsStyleless`
- `closeMenu`
- `pAssertMenuPresence`

You also could move the following I believe if you update it to pass in the expected label/style:
- `openMenu`
- `assertSingleCellStructureHasStyle`
- `assertMultipleCellsStructureHasStyle`
- `pAssertCheckmark`

- [ ] Remove test helper duplication

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderWidthTest.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderWidthTest.ts:39]: Same issue as the previous test. This never seems to be used, so please make sure to remove it or use it.

**Review by lnewson — APPROVED** [review]

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderStyleTest.ts
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderStyleTest.ts:39]: Because I copy-pasted it, not thinking.

File: modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts:254]: We're likely going to need using a proper onAction here later, and I'd prefer not to have to double the amount of subentry-generators when we do, so I'd prefer to keep this onAction-functionality.

Comment (lnewson) [reply modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts:254]: That's not what I meant sorry, I literally meant a function to do the exec command. eg:
```
const applyTableCellStyleAction = (style: string) => (item: Item) => {
  editor.execCommand('mceTableApplyCellStyle', false, { [style]: item.value });
};
```
and then that could be used like so:
```
generateItems(editor, tableCellBorderWidthsList, 'tablecellborderwidth', applyTableCellStyleAction('broder-width'))
```

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderWidthTest.ts
Comment (JamesToohey) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderWidthTest.ts:169]: I _think_ this can be declared outside the `it` blocks and used by the helper functions directly. So you don't need to pass it around

**Review by JamesToohey — APPROVED** [review]

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderWidthTest.ts
Comment (lnewson) [reply modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderWidthTest.ts:169]: I generally would advise against that, as while unlikely it's possible things could change between when the describe runs and the `it` block executes. It would be fine for this case, but if you were getting the editor body or something in an iframe it could have changed.

Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/test/ts/browser/ui/TableBorderWidthTest.ts]: Due to the restructuring this isn't really valid anymore.

**Review by lnewson — APPROVED** [review]
Really nice work on pulling out those helper test utilities @HAFRMO :trophy:

File: modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts:197]: Right, right, my bad. I may have been a bit too tired and jumped at the first thing I saw. My bad.

File: modules/tinymce/src/plugins/table/main/ts/api/Settings.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/api/Settings.ts:24]:
```suggestion

const defaultCellBorderWidth = Arr.range(5, (i) => {
```
There's a missing new line here, so just adding that.

File: modules/tinymce/src/plugins/table/main/ts/ui/UiUtils.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/UiUtils.ts:32]:
```suggestion

const applyTableCellStyle = <T extends Item>(editor: Editor, style: string) =>
```

File: modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts:197]:
```suggestion
    fetch: generateItemsCallback(
      editor,
      verticalAlignValues,
      'tablecellverticalalign',
      (item) => item.text,
      applyTableCellStyle(editor, 'vertical-align')
    ),
```

These lines are getting quite long now (well over the 160char max we normally try to keep within). Maybe we should "chop" it down like this? I'll add a suggestion for each, but your call 😄

File: modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts:205]:
```suggestion
    fetch: generateItemsCallback(
      editor,
      tableCellBorderWidthsList,
      'tablecellborderwidth',
      (item) => item.title,
      applyTableCellStyle(editor, 'border-width')
    ),
```

File: modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/Buttons.ts:213]:
```suggestion
    fetch: generateItemsCallback(
      editor,
      tableCellBorderStylesList,
      'tablecellborderstyle',
      (item) => item.title,
      applyTableCellStyle(editor, 'border-style')
    ),
```

File: modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts:240]:
```suggestion
    getSubmenuItems: () => generateItems(
      editor,
      verticalAlignValues,
      'tablecellverticalalign',
      (item) => item.text,
      applyTableCellStyle(editor, 'vertical-align')
    ),
```

File: modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts:247]:
```suggestion
    getSubmenuItems: () => generateItems(
      editor,
      tableCellBorderWidthsList,
      'tablecellborderwidth',
      (item) => item.title,
      applyTableCellStyle(editor, 'border-width')
    ),
```

File: modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts:255]:
```suggestion
    getSubmenuItems: () => generateItems(
      editor,
      tableCellBorderStylesList,
      'tablecellborderstyle',
      (item) => item.title,
      applyTableCellStyle(editor, 'border-style')
    ),
```

**Review by lnewson — APPROVED** [review]

**Comment by lnewson** [comment]
I've also re-requested reviews on this since there were some pretty large changes with the test deduplication.

**Review by JamesToohey — APPROVED** [review]

File: modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts
Comment (ltrouton) [inline modules/tinymce/src/plugins/table/main/ts/ui/MenuItems.ts:254]: Looks like an extra line here
```suggestion
```

**Review by ltrouton — APPROVED** [review]

**Review by ashfordneil — APPROVED** [review]

**Review by lnewson — CHANGES_REQUESTED** [review]
Sorry @HAFRMO but upon final reviews for :green_circle: today, we noticed an issue here so going to have to block this for now. Will move the discussion to slack.

**Review by lnewson — APPROVED** [review]

**Pull request merged**
